airgun/views/cloud_inventory.py

```diff
@@ -0,0 +1,162 @@
+from airgun.views.common import SatTab
+from airgun.widgets import Search
+
+
+class InventoryTab(SatTab):
+    """Cloud Inventory Upload Tab element.
+
+    This is lightweight subclass needed because Cloud Inventory Upload
+    tabs contain icons, and widgetastic_patternfly Tab looks for exact match.
+    """
+    TAB_LOCATOR = ParametrizedLocator(
+        './/ul[contains(@class, "nav-tabs")]/'
+        'li[./a[contains(normalize-space(.), {@tab_name|quote})]]'
+    )
+
+
+class InventoryBootstrapSwitch(Widget):
+    """ Checkbox-like Switch control, representing On and Off state. But with
+    fancy UI and without any <form> elements.
+    There's also BootstrapSwitch widget in widgetastic_patternfly, but we don't
+    inherit from it as it uses completely different HTML structure than this one
+    (it has underlying <input>).
+    """
+
+    ROOT = "//div[@class='auto_upload_switcher']/div"
+    ON_TOGGLE = ".//span[contains(@class, 'bootstrap-switch-handle-on')]"
+    OFF_TOGGLE = ".//span[contains(@class, 'bootstrap-switch-handle-off')]"
+
+    @property
+    def selected(self):
+        classes = self.browser.classes(self)
+        return 'bootstrap-switch-on' in classes
+
+    @property
+    def is_displayed(self):
+        return self.browser.is_displayed(locator=self.ROOT)
+
+    @property
+    def _clickable_el(self):
+        """ In automation, you need to click on exact toggle element to trigger action
+
+        Returns: selenium webelement
+        """
+        locator = self.ON_TOGGLE
+        if not self.selected:
+            locator = self.OFF_TOGGLE
+        return self.browser.element(locator=locator)
+
+    def fill(self, value):
+        value = bool(value)
+        current_value = self.selected
+        if value == current_value:
+            return False
+        else:
+            self.browser.click(self._clickable_el)
+            return True
+
+    def read(self):
+        return self.selected
+
+
+class InventoryItem(ListItem):
+    """Item related to one organization on Cloud Inventory Upload page."""
+    DESCRIPTION_LOCATOR = './/div[contains(@class, "pf-description")]'
+    STATUS_ELEMENTS = (
+        './/div[contains(@class, "status")][contains(@class, "container")]'
+        '/div[contains(@class, "item")]'
+    )
+
+    @View.nested
+    class generating(InventoryTab):
+        process = Text(".//div[contains(@class, 'tab-header')]/div[1]")
+        restart = Button('Restart')
+        terminal = Text(".//div[@class='terminal']")
+        scheduled_run = Text(".//div[contains(@class, 'scheduled_run')]")
+
+    @View.nested
+    class uploading(InventoryTab):
+        process = Text(".//div[contains(@class, 'tab-header')]/div[1]")
+        download_report = Button('contains', 'Download Report')
+        terminal = Text(".//div[@class='terminal']")
+
+    @property
+    def is_generating(self):
+        try:
+            self.parent_browser.selenium.find_element_by_xpath(f"{self.STATUS_ELEMENTS}[1]/div")
```

Comment:
why we are using the `self.parent_browser.selenium.find_element_by_xpath(f"{self.STATUS_ELEMENTS}[1]/div")` is it not possible to get that in any widget ?

Reply:
Unfortunately, no.

This xpath catches "spinner" ("loading wheel") element, that is only present when work is being done. I want immediate information whether element is present or not, as this function is used by `wait_for` to decide whether we can continue interacting with page, or not.

The way widgetastic is written, all functions I could possibly use (`is_displayed`, `wait_for_element`, `element`), under the hood will call `elements()` recursively. While [`elements()` can take `check_safe=False`](https://github.com/RedHatQE/widgetastic.core/blob/master/src/widgetastic/browser.py#L222) argument to force return value sooner, even on not fully loaded page, it drops this argument in one of recursive calls. In other words, even if I explicitly tell widgetastic to not call `ensure_page_safe()`, it will call it anyway.

And our `ensure_page_safe()` will wait until all spinners on page disappear. But will wait only 30 seconds, without a way for caller to change timeout.

In other words, I know this spinner is there and will remain for some time (in my testing, at least for a minute, but that depends on number of registered hosts), and I have my own wait_for with longer timeout to wait for it to disappear. But under the hood, widgetastic would call default `ensure_page_safe()` with default timeout, which would throw exception before process finishes.

Possible fixes could be:

* change our `ensure_page_safe` to not include spinners, but I have no idea how many existing tests depend on this behavior
* fix widgetastic
* circumvent widgetastic and reach out to Selenium directly

I decided last option will be quickest to implement. Especially considering that I will want that change backported to older Satellite versions, where I can't update widgetastic easily anyway.

Reply:
Do you think we need to raise an issue for this in Airgun for suture reference? Okay, I can understand it is a quick and easy fix for now.

Reply:
It's probably right thing to do, but so low on my priorities list, I don't see myself doing that.
6.7.z, 6.6.z and 6.5.z are not going to benefit from widgetastic core change anyway.

Reply:
okay